### PR TITLE
[SNOW-594195] Add correlationId to logging

### PR DESCRIPTION
**Pull request by sfc-gh-rcheng**
Problem: multiple instances of kafka connect log to same spot causing confusion during debugging. Issue came up during a the Fidelity issue linked in the jira.

**Changes:**

LoggingHandler
- is a layer between Kafka Connect and the logging framework 
- has a static correlationId that is updated on start()
- logs when a new logger is returned

Logging.java -> EnableLogging.java
- retains the original idea of Logging.java as an abstract class other classes can extend for easy logging methods

Logging.logMessage() -> Utils.formatLogMessage()
- increases logging consistency: moved the format log message into LoggingHandler so consumers don't need to format the message themselves
- name change: since logMessage doesn't log the message
- moved to Utils.java since there are instances we use it to format strings for not just logging purposes

### CODE REVIEW

**Comment by sfc-gh-rcheng** [comment]
apologies, this really should have been multiple prs. I've highlighted the important parts

File: src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnector.java
Comment (sfc-gh-rcheng) [inline src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnector.java:40]: This is how all the LOGGERs will be instantiated. Since they are static, they are created with no correlationId (it will be set later on)

File: src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnector.java
Comment (sfc-gh-rcheng) [inline src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnector.java:79]: This is where the correlationId will be set for all loggers. The goal is to have a new correlationId every time KC starts up

File: src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnector.java
Comment (sfc-gh-rcheng) [inline src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnector.java:121]: This probably could have been in a separate PR, but I refactored the logMessage function. It is now in Utils.java since we use it in places for not just logging. The LoggingHandler calls the new logMessage function, so these are removed

File: src/main/java/com/snowflake/kafka/connector/Utils.java
Comment (sfc-gh-rcheng) [inline src/main/java/com/snowflake/kafka/connector/Utils.java:628]: Equivalent to the previous LogMessage function in Logging.java

File: src/main/java/com/snowflake/kafka/connector/internal/EnableLogging.java
Comment (sfc-gh-rcheng) [inline src/main/java/com/snowflake/kafka/connector/internal/EnableLogging.java:4]: This replaces the Logging.java class

**Comment by sfc-gh-rcheng** [comment]
Confirmed that loggers created without a correlationId eventually log messages with a correlationId.

1. The static Utils logger was created without correlationId
`[2022-09-16 01:15:26,500] INFO [SF_KAFKA_CONNECTOR] Created loggerHandler for class: 'com.snowflake.kafka.connector.Utils' without a correlationId. (com.snowflake.kafka.connector.internal.LoggerHandler)`
2. Kafka connector starts with a correlationId
`[2022-09-16 01:15:53,216] INFO [SF_KAFKA_CONNECTOR] [0f9b9a0e-34a9-4c16-92ae-ee3f3bd99d89]SnowflakeSinkConnector:start (com.snowflake.kafka.connector.SnowflakeSinkConnector)`
3. All logs emitted from Utils after start have the correlationId
`[2022-09-16 01:15:53,623] DEBUG [SF_KAFKA_CONNECTOR] [0f9b9a0e-34a9-4c16-92ae-ee3f3bd99d89]generated stage name: SNOWFLAKE_KAFKA_CONNECTOR_SnowflakeSinkConnector_revi_STAGE_SnowflakeSink_revi (com.snowflake.kafka.connector.Utils)`

File: docker-setup/snowflake-kafka-docker/kafka-connect.yml
Comment (sfc-gh-tzhang) [inline docker-setup/snowflake-kafka-docker/kafka-connect.yml:2]: why we're removing these template now?

File: src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnector.java
Comment (sfc-gh-tzhang) [inline src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnector.java:22]: no star import please

File: src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnector.java
Comment (sfc-gh-tzhang) [reply src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnector.java:121]: just want to make sure there is no behavior changes with this refactor?

File: src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnectorConfig.java
Comment (sfc-gh-tzhang) [inline src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnectorConfig.java:116]:
```suggestion
  private static final LoggerHandler LOGGER = new LoggerHandler(SnowflakeSinkConnectorConfig.class.getName());
```

File: src/main/java/com/snowflake/kafka/connector/internal/EnableLogging.java
Comment (sfc-gh-tzhang) [inline src/main/java/com/snowflake/kafka/connector/internal/EnableLogging.java:1]: please add copyright

File: src/main/java/com/snowflake/kafka/connector/internal/LoggerHandler.java
Comment (sfc-gh-tzhang) [inline src/main/java/com/snowflake/kafka/connector/internal/LoggerHandler.java:19]: @sfc-gh-japatel would add correlationId at instance level be sufficient for the Fidelity case? IIRC, we had issues to figure out which partition or which thread was generating the log

**Review by sfc-gh-tzhang — COMMENTED** [review]
I'm not sure only adding the id at instance level would be sufficient, but this is a good start! We could add more things over time.

**Comment by sfc-gh-tzhang** [comment]
Please also add a PR description about what we're trying to do here

File: src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnector.java
Comment (sfc-gh-rcheng) [reply src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnector.java:121]: None of this logging change should affect the connector functionality. LoggingHandler is just an additional layer that keeps track of the correlationId and injects it into the log string

File: docker-setup/snowflake-kafka-docker/kafka-connect.yml
Comment (sfc-gh-rcheng) [reply docker-setup/snowflake-kafka-docker/kafka-connect.yml:2]: oops this is a git ignore mistake, ill add it back it

File: src/main/java/com/snowflake/kafka/connector/internal/EnableLogging.java
Comment (sfc-gh-rcheng) [reply src/main/java/com/snowflake/kafka/connector/internal/EnableLogging.java:1]: I also noticed other files missing copyrights. Is the goal to have every file contain the copyright at the beginning?

File: src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnector.java
Comment (sfc-gh-rcheng) [reply src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnector.java:22]: I think this is a refactoring thing, I'll try to run the google formatter

Comment (sfc-gh-rcheng) [reply src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnector.java:21]: it was an intellij thing, got rid of the * thing

File: src/main/java/com/snowflake/kafka/connector/internal/LoggerHandler.java
Comment (sfc-gh-rcheng) [reply src/main/java/com/snowflake/kafka/connector/internal/LoggerHandler.java]: @sfc-gh-tzhang to clarify, is there one thread per partition?

Unfortunately I don't know how Java treats static properties/objects with threads, but I assume we may not be able to add thread level correlationIds for static classes. For non-static threads, it would be really easy to append a short additional hash to each logger with EnableLogging.java. I can make that change if needed

File: src/test/java/com/snowflake/kafka/connector/internal/LoggerHandlerTest.java
Comment (sfc-gh-rcheng) [inline src/test/java/com/snowflake/kafka/connector/internal/LoggerHandlerTest.java:29]: can we test this? Is there a way to catch the logs being emitted, or a way to mock the Logger?

File: src/main/java/com/snowflake/kafka/connector/internal/EnableLogging.java
Comment (sfc-gh-tzhang) [reply src/main/java/com/snowflake/kafka/connector/internal/EnableLogging.java:18]: yes

**Review by sfc-gh-tzhang — APPROVED** [review]
LGTM, thanks! Regarding the testing, it would be good to use Mockito or Powermock, but I'm fine with some simple test cases.

File: src/main/java/com/snowflake/kafka/connector/internal/EnableLogging.java
Comment (sfc-gh-rcheng) [reply src/main/java/com/snowflake/kafka/connector/internal/EnableLogging.java:18]: created corresponding jira

File: src/main/java/com/snowflake/kafka/connector/internal/LoggerHandler.java
Comment (sfc-gh-rcheng) [reply src/main/java/com/snowflake/kafka/connector/internal/LoggerHandler.java]: Discussed offline, would like to add a task correlationId. jira: https://snowflakecomputing.atlassian.net/browse/SNOW-665587

File: src/main/java/com/snowflake/kafka/connector/SnowflakeSinkTask.java
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/SnowflakeSinkTask.java:242]: can you add which partitions are assigned to this instance of the task. Requires a some computation but would worth in debugging. 
Also lets correct the message 
```suggestion
        "SnowflakeSinkTask[ID:{}]:open, TopicPartition size: {}", this.id, partitions.size());
```

Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/SnowflakeSinkTask.java:242]: Can you add this in the below forEach loop too?
```suggestion
partitions.forEach(topicPartition -> Logging.logMessage("SnowflakeSinkTask[ID:{}] Assigned Partition:", topicPartition));
```

File: src/main/java/com/snowflake/kafka/connector/SnowflakeSinkTask.java
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/SnowflakeSinkTask.java:312]: nit: extra " + " not needed in between. (Probably some gfmt thing) 
Lets fix this.

File: src/main/java/com/snowflake/kafka/connector/internal/EnableLogging.java
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/EnableLogging.java:2]: nit: 2022

File: src/main/java/com/snowflake/kafka/connector/internal/LoggerHandler.java
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/LoggerHandler.java:8]: nit: Javadocs

File: src/main/java/com/snowflake/kafka/connector/internal/LoggerHandler.java
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/LoggerHandler.java:117]: nit: redundant?

File: src/main/java/com/snowflake/kafka/connector/internal/LoggerHandler.java
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/LoggerHandler.java:40]: great fallback! thanks

File: src/main/java/com/snowflake/kafka/connector/internal/LoggerHandler.java
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/LoggerHandler.java:12]: umm, ideally, since this is not static final, it shouldnt be all uper case. should we instead use `loggerCorrelationId`?

File: src/test/java/com/snowflake/kafka/connector/internal/LoggerHandlerTest.java
Comment (sfc-gh-japatel) [reply src/test/java/com/snowflake/kafka/connector/internal/LoggerHandlerTest.java:29]: https://stackoverflow.com/a/1828268

**Comment by sfc-gh-japatel** [comment]
> Confirmed that loggers created without a correlationId eventually log messages with a correlationId.
> 
> 1. The static Utils logger was created without correlationId
>    `[2022-09-16 01:15:26,500] INFO [SF_KAFKA_CONNECTOR] Created loggerHandler for class: 'com.snowflake.kafka.connector.Utils' without a correlationId. (com.snowflake.kafka.connector.internal.LoggerHandler)`
> 2. Kafka connector starts with a correlationId
>    `[2022-09-16 01:15:53,216] INFO [SF_KAFKA_CONNECTOR] [0f9b9a0e-34a9-4c16-92ae-ee3f3bd99d89]SnowflakeSinkConnector:start (com.snowflake.kafka.connector.SnowflakeSinkConnector)`
> 3. All logs emitted from Utils after start have the correlationId
>    `[2022-09-16 01:15:53,623] DEBUG [SF_KAFKA_CONNECTOR] [0f9b9a0e-34a9-4c16-92ae-ee3f3bd99d89]generated stage name: SNOWFLAKE_KAFKA_CONNECTOR_SnowflakeSinkConnector_revi_STAGE_SnowflakeSink_revi (com.snowflake.kafka.connector.Utils)`

Should there be an extra space `[corr_id] here` instead of `[corr_id]here` ?

File: src/main/java/com/snowflake/kafka/connector/internal/LoggerHandler.java
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/LoggerHandler.java:121]: May be?
```suggestion
    return isCorrelationIdValid() ? "[" + LOGGER_CORRELATION_ID.toString() + "] " : "";
```

**Review by sfc-gh-japatel — COMMENTED** [review]
overall looks great. Thanks Revi! Will wait for your comments. [No major concerns]

**Comment by sfc-gh-japatel** [comment]
Can we somehow also get this correlation id in our telemetry so that we can check uuid in snowhouse? Dont need to do this in this PR but lets track it.

File: src/test/java/com/snowflake/kafka/connector/internal/LoggerHandlerTest.java
Comment (sfc-gh-rcheng) [reply src/test/java/com/snowflake/kafka/connector/internal/LoggerHandlerTest.java]: thanks! added testing

File: src/main/java/com/snowflake/kafka/connector/SnowflakeSinkTask.java
Comment (sfc-gh-rcheng) [reply src/main/java/com/snowflake/kafka/connector/SnowflakeSinkTask.java:242]: This is out of scope for this pr i believe, I'll add this in a follow up one for the debugging scenario. created jira tracking this: https://snowflakecomputing.atlassian.net/browse/SNOW-665898

File: src/main/java/com/snowflake/kafka/connector/internal/LoggerHandler.java
Comment (sfc-gh-rcheng) [reply src/main/java/com/snowflake/kafka/connector/internal/LoggerHandler.java:124]: Good point, removed the null check

File: src/main/java/com/snowflake/kafka/connector/internal/LoggerHandler.java
Comment (sfc-gh-rcheng) [reply src/main/java/com/snowflake/kafka/connector/internal/LoggerHandler.java]: yes! made the change thanks for catching that

File: src/main/java/com/snowflake/kafka/connector/SnowflakeSinkTask.java
Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/SnowflakeSinkTask.java:243]: Umm, okay. I think its fine to just add it here since it is only one more line but I will leave it up to you!

**Review by sfc-gh-japatel — APPROVED** [review]
lgtm, thanks for adding tests! Looks complete. 
Lets make sure the gfmt passes before you merge.

**Pull request merged**
